footnote reverse link

ldXiao opened the pull request:
在每处footnote调用处添加一个`footnote-origin:\thefootnote`的label，在`printfoot`处引用此label

生成效果如下

<img width="516" alt="image" src="https://github.com/banana-space/btex/assets/32606147/86fe2baf-60d2-43c3-89ec-2fddec5c9fd9">

点击`^`将会返回脚标源头

ldXiao committed:
footnote reverse link

ldXiao committed:
decouple footnoteorigin from footnotemakr and footnotetext

abccsss commented:
应解决以下问题：

1. `^` 之后应带有空格。

2. 当一个脚注有多个链接时不能正常输出。例如：
    ```latex
    \footnotemark\footnotemark\footnotetext{1}
    \printfootnotes
    ```

ldXiao committed:
code clean up

ldXiao commented:
修改过后之后只有`\footnote`会产生反向链接，裸`\footnotemark\footnotetext{....}`不会产生反向链接

abccsss commented:
问题：

3. 所有 `^` 链接都指向同一个脚注。

4. 生成的 `^` 目前是
    ```html
    <span><a href="#o1"><sup>^</sup></a></span><sup style="white-space:pre-wrap"> </sup>
    ```
    应该是
    ```html
    <span><a href="#o1">^</a> </span>
    ```
    注意不要使用 `<sup>` (否则符号过小)，不必使用 `\ ` 产生强制空格。

5. 多个脚注同一标号的功能也应该实现，但目前不是必须的。

ldXiao committed:
temp working

ldXiao committed:
same counter but makes more sense

ldXiao committed:
remove explicit whitespace

ldXiao commented:
3，4已经修正

5. 现有代码可以实现`多个脚注同一标号的功能`，但是正确用法是`x\footnotemark ... y\footnotemark ... z\footnote{text}`

然后反向链接会指向最后一处`\footnote`

更完善的解决方案有点费脑子，我再想想

ldXiao commented:
```latex
\def\@footnoteoriginrefs{}

\def\@footnotemark{%
    \@inc{g.ctr-footnote}%
    {%
        \@settrue{text-sup}%
        \@@set{ref-key}{footnote:{g.ctr-footnote}}%
        \@ref{\thefootnote}%
        \@footnoteorigin%
        \@patch\@footnoteoriginrefs{\@footnoteoriginref}
    }%
    \@dec{g.ctr-footnote}%
}

\def\@printfootnoteitem[#1][#2]{%
    \@@set{ref-prefix}{f}%
    \@item[#2.]%
    \@settrue{text-sup}%
    \@@def\@currentlabel{#2}%
    \@@bmk%
    \@@label{footnote:#1}%
    \@setfalse{text-sup}%
}

\def\@footnotetext#1{%
    \@inc{g.ctr-actualfootnote}%
    \@patch\@printfootnotes{\@printfootnoteitem[}%
    \@defexpand\@patch\@printfootnotes{\theactualfootnote}%
    \@patch\@printfootnotes{][}%
    \@defexpand\@patch\@printfootnotes{\thefootnote}%
    \@patch\@printfootnotes{]#1}%
}

\let\footnotemark\@footnotemark

\def\footnotetext#1{%
    \@inc{g.ctr-footnote}%
    \@footnotetext{
        \@@get{footnote-origin-refs}#1
    }%
    \def\@footnoteoriginrefs{}
}

\pdef\@footnoteorigin{%
    \@inc{g.ctr-footnoteorigin}
    \@@def\@currentlabel{^}%
    \@@set{ref-prefix}{o}% 
    \@@bmk%
    \@@label{footnote-origin:{g.ctr-footnoteorigin}}%
}

\pdef\@footnoteoriginref{%
    {
        \@inc{g.ctr-footnoteoriginref}
        \@@set{ref-key}{footnote-origin:{g.ctr-footnoteoriginref}}%
        \@ref{\thefootnoteorigin}%
    }
}
```
假如能够用一个collection 把 footnoteoriginref存起来，每次`\footnotemark`调用`\@patch\@footnoteoriginrefcollection{\@footnoteoriginref}`，并且在`printfootnotes`里面每行调用`\@footnoteoriginrefcollection`的*过程中的值*就可以实现下图这样的效果

<img width="212" alt="image" src="https://github.com/banana-space/btex/assets/32606147/f15b92b8-4f43-4bdf-9967-3ec70fd0f447">

然而`\@footnoteoriginrefcollection`只能存储命令被call的时候的值，而`printfootnotes`在结尾也就是这些中间状态全部都丢失了

暂时没有想到好的方法

ldXiao committed:
enable separate footnotemark reverse link

ldXiao committed:
code clean up

ldXiao commented:
到这一步已经解决了多个footnotemark 各自reverse link
可以实现以下效果

![image](https://github.com/banana-space/btex/assets/32606147/f216449e-90ec-40b1-9cca-b26030fc8aa6)

![image](https://github.com/banana-space/btex/assets/32606147/2395836d-2c22-41a4-ad3e-387454cbd843)

每个反向链接`^`都会跳转到对应的footnotemark

类似的机制也可以做多处引用参考文献的反向跳转

假如你对实现的效果没有意见，那么稍微清理一下也是可以合并的

abccsss commented:
需要修改的问题：

6. 问题 4 未完全修正，需要将 `<sup>^</sup>` 改成 `^` (或者说明不这么做的理由)。

7. 表格中不能正常插入脚注。例如
    ```latex
    \footnotemark\footnotemark
    \begin{tabular}{cc}
        1\footnotetext{first} &
        2\footnote{second}
    \end{tabular}

    \printfootnotes
    ```
    会把第二个脚注的文字写成 `4 5 second`，并产生一个编译错误。原来的版本没有此问题。

8. 目前生成的脚注 HTML 是
    ```html
    <span><a href="#f1"><sup>1</sup></a></span><sup> </sup><span id="o1"></span>
    ```
    脚注标记后面有个多余的空格 `<sup> </sup>`，需要去掉，以符合某些排版需求，例如脚注后面有标点时就不能有空格。

ldXiao committed:
resolve following issues:
use % signs to clear superficial paragraph
use \gdef\@footnoteorigincollection{}%
to make it work in tabular cell

ldXiao commented:
你现在再试一下

ldXiao committed:
remove superficial space
